src/app/BufferedReadCallback.cpp

```diff
@@ -0,0 +1,239 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "lib/core/CHIPTLV.h"
+#include "lib/core/CHIPTLVTags.h"
+#include "lib/core/CHIPTLVTypes.h"
+#include "protocols/interaction_model/Constants.h"
+#include "system/SystemPacketBuffer.h"
+#include "system/TLVPacketBufferBackingStore.h"
+#include <app/BufferedReadCallback.h>
+#include <app/InteractionModelEngine.h>
+
+namespace chip {
+namespace app {
+
+void BufferedReadCallback::OnReportBegin()
+{
+    mCallback.OnReportBegin();
+}
+
+void BufferedReadCallback::OnReportEnd()
+{
+    DispatchBufferedData(nullptr, mBufferedPath, StatusIB(), true);
+    mCallback.OnReportEnd();
+}
+
+CHIP_ERROR BufferedReadCallback::GenerateListTLV(System::PacketBufferTLVReader & aReader)
+{
+    System::PacketBufferHandle handle = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSizeWithoutReserve, 0);
+    System::PacketBufferTLVWriter writer;
+    CHIP_ERROR err;
+    TLV::TLVType outerType;
+
+    //
+    // Initialize the writer, but with chaining enabled, since we'll very likely
+    // exceed the size of a single packet buffer when reconstituing the entire list
+    //
+    writer.Init(std::move(handle), /* useChainedBuffers */ true);
+
+    err = writer.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Array, outerType);
+    SuccessOrExit(err);
+
+    for (auto & bufHandle : mBufferedList)
+    {
+        System::PacketBufferTLVReader reader;
+
+        reader.Init(std::move(bufHandle));
+
+        err = reader.Next();
+        SuccessOrExit(err);
+
+        err = writer.CopyElement(TLV::AnonymousTag, reader);
+        SuccessOrExit(err);
+    }
+
+    err = writer.EndContainer(outerType);
+    SuccessOrExit(err);
+
+    writer.Finalize(&handle);
+    aReader.Init(std::move(handle));
+
+exit:
```

Comment:
since this exit block has no cleanup code, could we use returns instead? goto exit pattern forces C-style variable declarations which are not always convenient.

Same comment throughout this file.